File: sdks/python/src/graphrec_sdk/ecommerce/session.py
"""Recommendation widgets with automatic feedback bookkeeping.

::

    from graphrec_sdk.ecommerce import RecommendationSession

    recs_ui = RecommendationSession(client)

    # product page: recommend and record the impression in one call
    recs = recs_ui.recommend(user_id="customer-42", top_n=8, exclude_product_ids=["sku-123"])

    # later, when the shopper clicks / buys a recommended item
    recs_ui.click(recs, "sku-777")
    recs_ui.convert(recs, "sku-777", value="49.90")

Clicks are linked to the impression automatically and positions are looked up
from the original response. Store ``recs.request_id`` (e.g. in the page or the
cart line) if the click happens in another process, then call
``client.feedback.click(request_id, product_id, position=...)`` directly.
"""

from __future__ import annotations

import threading
from collections import OrderedDict
from decimal import Decimal
from typing import TYPE_CHECKING, Any, Dict, Mapping, Optional, Sequence, Union

from ..models.recommendations import FeedbackReceipt, Recommendations

if TYPE_CHECKING:
    from .._client import AsyncGraphRec, GraphRec

__all__ = ["AsyncRecommendationSession", "RecommendationSession"]

Money = Union[Decimal, float, int, str]


class _ImpressionIndex:
    """Bounded map ``request_id -> impression event_id``."""

    def __init__(self, capacity: int) -> None:
        self.capacity = max(capacity, 1)
        self._data: OrderedDict[str, str] = OrderedDict()
        self._lock = threading.Lock()

    def put(self, request_id: str, event_id: str) -> None:
        with self._lock:
            self._data[request_id] = event_id
            self._data.move_to_end(request_id)
            while len(self._data) > self.capacity:
                self._data.popitem(last=False)

    def get(self, request_id: str) -> Optional[str]:
        with self._lock:
            return self._data.get(request_id)


def _request_args(
    user_id: Optional[str],
    top_n: int,
    context: Optional[Mapping[str, Any]],
    exclude_product_ids: Optional[Sequence[str]],
) -> Dict[str, Any]:
    return {
        "user_id": user_id,
        "top_n": top_n,
        "context": context,
        "exclude_product_ids": exclude_product_ids,
    }


class RecommendationSession:
    """Fetch recommendations and send impression/click/conversion feedback consistently."""

    def __init__(
        self, client: GraphRec, *, auto_impression: bool = True, remember: int = 1_000
    ) -> None:
        self._client = client
        self.auto_impression = auto_impression
        self._impressions = _ImpressionIndex(remember)

    def recommend(
        self,
        *,
        user_id: Optional[str] = None,
        session_id: Optional[str] = None,
        recent_product_ids: Optional[Sequence[str]] = None,
        top_n: int = 10,
        context: Optional[Mapping[str, Any]] = None,
        exclude_product_ids: Optional[Sequence[str]] = None,
    ) -> Recommendations:
        """Personalized results for ``user_id``, or session-based when ``session_id`` is given."""

        args = _request_args(user_id, top_n, context, exclude_product_ids)
        if session_id is not None:
            recs = self._client.recommendations.for_session(
                session_id, recent_product_ids=recent_product_ids, **args
            )
        else:
            recs = self._client.recommendations.get(**args)
        if self.auto_impression and recs.items:
            self.impression(recs)
        return recs

    def impression(self, recs: Recommendations, **options: Any) -> FeedbackReceipt:
        receipt = self._client.feedback.impression(recs, **options)
        self._impressions.put(recs.request_id, receipt.event_id)
        return receipt

    def click(self, recs: Recommendations, product_id: str, **options: Any) -> FeedbackReceipt:
        options.setdefault("impression_event_id", self._impressions.get(recs.request_id))
        return self._client.feedback.click(recs, product_id, **options)

    def convert(
        self,
        recs: Recommendations,
        product_id: str,
        *,
        value: Optional[Money] = None,
        **options: Any,
    ) -> FeedbackReceipt:
        return self._client.feedback.conversion(recs, product_id, value=value, **options)


class AsyncRecommendationSession:
    """Async variant of :class:`RecommendationSession`."""

    def __init__(
        self, client: AsyncGraphRec, *, auto_impression: bool = True, remember: int = 1_000
    ) -> None:
        self._client = client
        self.auto_impression = auto_impression
        self._impressions = _ImpressionIndex(remember)

    async def recommend(
        self,
        *,
        user_id: Optional[str] = None,
        session_id: Optional[str] = None,
        recent_product_ids: Optional[Sequence[str]] = None,
        top_n: int = 10,
        context: Optional[Mapping[str, Any]] = None,
        exclude_product_ids: Optional[Sequence[str]] = None,
    ) -> Recommendations:
        args = _request_args(user_id, top_n, context, exclude_product_ids)
        if session_id is not None:
            recs = await self._client.recommendations.for_session(
                session_id, recent_product_ids=recent_product_ids, **args
            )
        else:
            recs = await self._client.recommendations.get(**args)
        if self.auto_impression and recs.items:
            await self.impression(recs)
        return recs

    async def impression(self, recs: Recommendations, **options: Any) -> FeedbackReceipt:
        receipt = await self._client.feedback.impression(recs, **options)
        self._impressions.put(recs.request_id, receipt.event_id)
        return receipt

    async def click(
        self, recs: Recommendations, product_id: str, **options: Any
    ) -> FeedbackReceipt:
        options.setdefault("impression_event_id", self._impressions.get(recs.request_id))
        return await self._client.feedback.click(recs, product_id, **options)

    async def convert(
        self,
        recs: Recommendations,
        product_id: str,
        *,
        value: Optional[Money] = None,
        **options: Any,
    ) -> FeedbackReceipt:
        return await self._client.feedback.conversion(recs, product_id, value=value, **options)

File: sdks/typescript/src/ecommerce/session.ts
/**
 * Fetch recommendations and send impression, click and conversion feedback
 * consistently. Impressions are recorded automatically and linked to later
 * clicks by `request_id`.
 *
 *     const widget = new RecommendationSession(client);
 *     const recs = await widget.recommend({ userId: "customer-42", topN: 6 });
 *     await widget.click(recs, recs.items[0].external_product_id);
 *     await widget.convert(recs, recs.items[0].external_product_id, { value: "59.00" });
 */
import type { GraphRec } from "../client.js";
import type { ClickOptions, ConversionOptions, ImpressionOptions, RecommendationOptions } from "../resources/recommendations.js";
import type { FeedbackReceipt, Recommendations } from "../types.js";

export interface RecommendOptions extends RecommendationOptions {
  /** Session-based recommendations for anonymous shoppers. */
  sessionId?: string;
  recentProductIds?: readonly string[];
}

export interface RecommendationSessionOptions {
  /** Record an impression for every non-empty result (default `true`). */
  autoImpression?: boolean;
  /** How many `request_id -> impression event_id` links to remember (default 1000). */
  remember?: number;
}

/** Bounded insertion-ordered map. */
class ImpressionIndex {
  private readonly data = new Map<string, string>();
  constructor(private readonly capacity: number) {}
  put(requestId: string, eventId: string): void {
    this.data.delete(requestId);
    this.data.set(requestId, eventId);
    while (this.data.size > this.capacity) {
      const oldest = this.data.keys().next().value;
      if (oldest === undefined) break;
      this.data.delete(oldest);
    }
  }
  get(requestId: string): string | undefined {
    return this.data.get(requestId);
  }
}

export class RecommendationSession {
  readonly autoImpression: boolean;
  private readonly impressions: ImpressionIndex;

  constructor(private readonly client: GraphRec, options: RecommendationSessionOptions = {}) {
    this.autoImpression = options.autoImpression ?? true;
    this.impressions = new ImpressionIndex(Math.max(1, options.remember ?? 1000));
  }

  /** Personalized results for `userId`, or session-based when `sessionId` is given. */
  async recommend(options: RecommendOptions = {}): Promise<Recommendations> {
    const { sessionId, recentProductIds, ...rest } = options;
    const recs = sessionId !== undefined ? await this.client.recommendations.forSession({ ...rest, sessionId, recentProductIds }) : await this.client.recommendations.get(rest);
    if (this.autoImpression && recs.items.length) await this.impression(recs);
    return recs;
  }

  async impression(recs: Recommendations, options: ImpressionOptions = {}): Promise<FeedbackReceipt> {
    const receipt = await this.client.feedback.impression(recs, options);
    this.impressions.put(recs.request_id, receipt.event_id);
    return receipt;
  }

  async click(recs: Recommendations, productId: string, options: ClickOptions = {}): Promise<FeedbackReceipt> {
    return this.client.feedback.click(recs, productId, { impressionEventId: this.impressions.get(recs.request_id), ...options });
  }

  async convert(recs: Recommendations, productId: string, options: ConversionOptions = {}): Promise<FeedbackReceipt> {
    return this.client.feedback.conversion(recs, productId, options);
  }
}
